### DISCUSSION TITLE
`JsonPropertyInfo.AttributeProvider` is null when using `JsonSerializerContext`

### DISCUSSION THREAD

**Discussion by Dean-NC** [dotnet/runtime#96532] [General] [answered]
I use JSON source gen. I also need to use a modifier, and if the property has a certain custom attribute then change `JsonPropertyInfo.Name` according to a property on that attribute. But when using `IJsonTypeInfoResolver.WithAddedModifier()`, `JsonPropertyInfo.AttributeProvider` is null.

```
[JsonSourceGenerationOptions(
    GenerationMode = JsonSourceGenerationMode.Metadata,
    PropertyNamingPolicy = JsonKnownNamingPolicy.CamelCase,
    PropertyNameCaseInsensitive = true)]
[JsonSerializable(typeof(IEnumerable<Company>))]
internal sealed partial class DeserializeJsonContext : JsonSerializerContext
{}

public static readonly JsonSerializerOptions MyDeserializeOptions = new()
{
    TypeInfoResolver = DeserializeJsonContext.Default.WithAddedModifier(MyModifer)
};

public static void MyModifier(JsonTypeInfo typeInfo)
{
    if (typeInfo.Kind != JsonTypeInfoKind.Object)
        return;

    Type customAttributeType = typeof(MyAttribute);
    foreach (var property in typeInfo.Properties)
    {
        if (property.AttributeProvider?.GetCustomAttributes(customAttributeType, true) is { } propertyAttributes && propertyAttributes.Length > 0)
        {
             if (propertyAttributes.FirstOrDefault() is MyAttribute attribute && !string.IsNullOrEmpty(attribute.Name))
             {
                 property.Name = attribute.Name;
             } 
        }
    }
}
```

I'm guessing it's because source-gen knows it doesn't need custom attributes it doesn't know about so it doesn't bother letting us access `AttributeProvider`, or is something else wrong?

Thanks much.

**Comment by eiriktsarpalis** [comment, marked as answer]
> I'm guessing it's because source-gen knows it doesn't need custom attributes it doesn't know about so it doesn't bother letting us access AttributeProvider, or is something else wrong?

That's correct, the source generator doesn't need those so it also doesn't attempt to populate them. The source generator could theoretically generate code that queries the relevant `MemberInfo` and populates the `AttributeProvider` property, however that isn't pay-for-play and it would negatively impact startup time for applications that don't apply contract customization.

My recommendation would be to insert an equivalent querying logic in your customization code, roughly:
```C#
property.AttributeProvider ??= typeInfo.Type.GetMember(property.Name, MemberTypes.Property | MemberTypes.Field, <relevant binding flags>);
```

**Reply by Dean-NC** [reply]
@eiriktsarpalis Thank you for your reply and suggestion. I had to change it to `typeInfo.Type.GetProperty` but it works great. `GetProperty()` runs fast, but still, it's good that the result of the modifier is cached so it runs once for each type.

**Reply by eiriktsarpalis** [reply]
Apropos, one potential issue with my proposed solution is that `JsonPropertyInfo.Name` might not necessarily reflect the actual name of the .NET member (e.g. because a `JsonPropertyName` attribute has been applied). Assuming you control the types you're applying the contract customizer to that shouldn't be an issue though.

**Comment by xiaoyuvax** [comment]
neither works with NativeAOT, do u know how to solve this?  seems `JsonTypeInfo.Type` should be annotated.


> warning IL2075: 'this' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicConstructors', 'DynamicallyAccessedMemberTypes.PublicMethods', 'DynamicallyAccessedMemberTypes.PublicFields', 'DynamicallyAccessedMemberTypes.PublicNestedTypes', 'DynamicallyAccessedMemberTypes.PublicProperties', 'DynamicallyAccessedMemberTypes.PublicEvents' in call to 'System.Type.GetMember(String, BindingFlags)'. The return value of method 'System.Text.Json.Serialization.Metadata.JsonTypeInfo.Type.get' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.

in code:
```
private static void DetectCustomAttribute([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] JsonTypeInfo typeInfo)
{
    if (typeInfo.Kind == JsonTypeInfoKind.Object)
        foreach (var pi in
            from JsonPropertyInfo propertyInfo in typeInfo.Properties
            from MemberInfo mi in typeInfo.Type.GetMember(propertyInfo.Name, BindingFlags.Public | BindingFlags.GetProperty)
            where  mi.Name== propertyInfo.Name && mi.GetCustomAttribute<CustomAttribute>() == null 
            select propertyInfo)
        {
            pi.Get = null;
            pi.Set = null;
        }
}
```
btw, is this legal???

`property.AttributeProvider ??= typeInfo.Type.GetMember(property.Name, MemberTypes.Property | MemberTypes.Field, <relevant binding flags>);`


but after i compiled this with NativeAot, it does work, so do the Warnings really matter? 
still, `JsonTypeInfo.Type` better be annotated as to eliminate the warning.
finally, i adopted GetProperties() instead of GetMember().

**Reply by eiriktsarpalis** [reply]
Using reflection to query members is theoretically speaking trim-unsafe, however assuming you're using a source generator you should be safe in this case since all properties have already been statically referenced by generated code. You could try to just suppress the warning:
```C#
[UnconditionalSuppressMessage("All members have been statically referenced by the source generator", "IL2075")]
private static void DetectCustomAttribute(JsonTypeInfo typeInfo)
```
